Dockerfile

```diff
@@ -50,73 +50,71 @@ RUN apt-get update -qq && \
       tzdata
 
 # Install application gems
+COPY bin ./bin
+
 COPY Gemfile Gemfile.lock ./
 RUN gem update --system && gem install bundler
 
-COPY ./bin/bundle ./bin/bundle
-RUN ./bin/bundle install && \
-    rm -rf ~/.bundle/ "${BUNDLE_PATH}"/ruby/*/cache "${BUNDLE_PATH}"/ruby/*/bundler/gems/*/.git
+RUN bundle config set --local path /rails/.cache/bundle
+
+# Use a cache mount to reuse old gems
+RUN --mount=type=cache,sharing=private,target=/rails/.cache/bundle \
+  mkdir -p vendor && \
+  bundle install && \
+  cp -ar /rails/.cache/bundle vendor/bundle && \
+  bundle config set path vendor/bundle
 
 # Install node dependencies
 COPY package.json yarn.lock .yarnrc.yml ./
-COPY ./bin/yarn ./bin/yarn
 RUN ./bin/yarn install --immutable
 
 COPY . .
 
-RUN ASSETS_COMPILATION=true SECRET_KEY_BASE=1 ./bin/bundle exec i18n export
-RUN ASSETS_COMPILATION=true SECRET_KEY_BASE=1 ./bin/rake assets:precompile
+RUN ASSETS_COMPILATION=true SECRET_KEY_BASE=1 RAILS_MAX_THREADS=4 NODE_OPTIONS="--max_old_space_size=4096" ./bin/bundle exec i18n export
+RUN --mount=type=cache,uid=1000,target=/rails/tmp/cache ASSETS_COMPILATION=true SECRET_KEY_BASE=1 RAILS_MAX_THREADS=4 NODE_OPTIONS="--max_old_space_size=4096" ./bin/rake assets:precompile
 
 RUN rm -rf node_modules
 
 FROM base AS runtime
 
+# Install fonts for rendering PDFs (mostly competition summary PDFs)
+# dejavu = Hebrew, Arabic, Greek
+# unfonts-core = Korean
+# wqy-modern = Chinese
+# ipafont = Japanese
+# thai-tlwg = Thai (as the name suggests)
+# lmodern = Random accents and special symbols for Latin script
+
 RUN apt-get update -qq && \
     apt-get install --no-install-recommends -y \
       mariadb-client \
       zip \
-      python-is-python3
+      python-is-python3 \
+      fonts-dejavu \
+      fonts-unfonts-core \
+      fonts-wqy-microhei \
+      fonts-ipafont \
+      fonts-thai-tlwg \
+      fonts-lmodern
+
+RUN useradd rails --create-home --shell /bin/bash
+USER rails:rails
 
 # Copy built artifacts: gems, application
 COPY --from=build /rails .
 
-# Run and own only the runtime files as a non-root user for security
-RUN useradd rails --create-home --shell /bin/bash && \
-    chown -R rails:rails vendor db log tmp public app pids
```

Comment:
I'm assuming that `chown` is not necessary anymore because we're switching to the `rails` user so early now that all `COPY` operations are creating the files as `rails` anyways. Correct?

Reply:
exactly!